reconcile/utils/terraform_config.py

```diff
@@ -0,0 +1,104 @@
+from typing import Iterable, Optional, Protocol
+from reconcile.utils.external_resource_spec import (
+    ExternalResourceSpecInventory,
+    ExternalResourceSpec,
+)
+
+from reconcile.utils.ocm import OCMMap
+
+
+class UnknownProvisionProviderError(Exception):
+    pass
+
+
+class ProvisionProviderAlreadyRegisteredError(Exception):
+    pass
+
+
+class TerraformClientProtocol(Protocol):
+    def dump(
+        self,
+        print_to_file: Optional[str] = None,
+        existing_dirs: Optional[dict[str, str]] = None,
+    ) -> dict[str, str]:
+        ...
+
+    def init_populate_specs(
+        self, specs: Iterable[ExternalResourceSpec], account_name: Optional[str]
+    ) -> None:
+        ...
+
+    def populate_resources(
+        self, spec: ExternalResourceSpec, ocm_map: Optional[OCMMap] = None
+    ):
+        ...
+
+
+class TerraformConfigProvider:
+    """
+    At a high-level, this class is responsible for generating Terraform configuration in
+    JSON format from app-interface schemas/openshift/external-resource-1.yml objects.
+    Usage example (mostly to demonstrate API):
+
+    cp = TerraformConfigProvider()
+    cp.register_terraform_client("aws", terrascript_aws_client)
+    cp.init_spec_inventory(specs, provision_provider, provisioner_name)
+    cp.populate_resources(ocm_map=ocm_map)
+    cp.dump(print_to_file, existing_dirs=working_dirs)
+    """
+
+    def __init__(self):
+        self.resource_spec_inventory: ExternalResourceSpecInventory = {}
+        self._terraform_clients: dict[str, TerraformClientProtocol] = {}
+
+    def register_terraform_client(
+        self, provision_provider: str, client: TerraformClientProtocol
+    ):
+        if not self._terraform_clients.get(provision_provider):
+            self._terraform_clients[provision_provider] = client
+        else:
+            raise ProvisionProviderAlreadyRegisteredError(
+                f"There can only be one client for each provision provider, {provision_provider} was already registered"
+            )
+
+    def dump(
```

Comment:
Terrascript or the Terraform CDK should be the entry point for JSON generation more rather than the providers. You can see this in the `__str__()` method of `terrascript.Terrascript`. With the Terraform CDK it would probably be the `synth` sub-command (might be other options directly in code as well).

There is probably some other shared logic in `dump()`, but this method isn't only used by `terraform-resources`, so it would require some other refactoring. I may see if there are any other refactoring wins to remove all direct references to `TerrascriptClient`, but it's not a primary goal of this PR at the moment.

Reply:
> There is probably some other shared logic in `dump()`, but this method isn't only used by `terraform-resources`, so it would require some other refactoring. I may see if there are any other refactoring wins to remove all direct references to `TerrascriptClient`, but it's not a primary goal of this PR at the moment.

Yeah, the `TerrascriptClient` is directly coupled with other modules. We could alternatively add another layer in between with the contract you have in `TerraformClientProvider` and through composition inject different config providers (TerrascriptClient as it exists and in future anything that implements `TerraformClientProtocol`). That could theoretically decouple provider from configuration logic and allow flexibility of choosing any `TerraformClientProtocol`. Did you consider this approach and were there significant hurdles with it?